src/main/scala/org/fortysevendeg/sparksftp/ReadingSFTPHadoopApp.scala

```diff
@@ -3,44 +3,77 @@ package org.fortysevendeg.sparksftp
+import org.fortysevendeg.sparksftp.common.HiveUserData._
+import org.fortysevendeg.sparksftp.common.HiveUserData
 import org.fortysevendeg.sparksftp.common.SparkUtils._
 import org.fortysevendeg.sparksftp.config.model.configs
-import org.fortysevendeg.sparksftp.config.model.configs.ReadingSFTPConfig
+import org.fortysevendeg.sparksftp.config.model.configs.{ReadingSFTPConfig, SFTPConfig}
 import org.training.trainingbot.config.ConfigLoader
 
+case class Salary(name: String, salary: Long)
+
 object ReadingSFTPHadoopApp extends IOApp {
 
   def setupConfig: IO[ReadingSFTPConfig] =
     ConfigLoader[IO]
       .loadConfig[ReadingSFTPConfig]
 
-  def run(args: List[String]): IO[ExitCode] =
-    for {
-      // Read the application parameters and prepare the spark session.
-      config: ReadingSFTPConfig <- setupConfig
-      defaultSparkConf: SparkConf = createSparkConfWithSFTPSupport(config)
+  /**
+   * When preparing the SparkConf, we could also use:
+   * .set("spark.master", "local")
+   * .set("spark.master", "local[*]")
+   *  to run the application locally, for example, through "sbt run"
+   */
+  def run(args: List[String]): IO[ExitCode] = {
+    def createSparkSession(config: ReadingSFTPConfig): IO[SparkSession] = IO {
+      val defaultSparkConf: SparkConf = createSparkConfWithSFTPSupport(config)
         .set("fs.sftp.impl", "org.apache.hadoop.fs.sftpwithseek.SFTPFileSystem")
         .set("fs.sftp.impl.disable.cache", "true")
-      sparkSession = getSparkSessionWithHive(defaultSparkConf)
-
-      sftpConfig = configs.SFTPConfig
-        .configFromContextProperties(sparkSession.sparkContext, config.sftp)
+      getSparkSessionWithHive(defaultSparkConf)
+    }
 
-      // Read the source files from SFTP into dataframes
-      usersPath    = s"sftp://${sftpConfig.sftpUser}:${sftpConfig.sftpPass}@${sftpConfig.sftpHost}" + s":${sftpConfig.sftpUserPath}"
-      salariesPath = s"sftp://${sftpConfig.sftpUser}:${sftpConfig.sftpPass}@${sftpConfig.sftpHost}" + s":${sftpConfig.sftpSalaryPath}"
+    def userPath(sftpConfig: SFTPConfig): String =
+      s"sftp://${sftpConfig.sftpUser}:${sftpConfig.sftpPass}@${sftpConfig.sftpHost}" + s":${sftpConfig.sftpUserPath}"
 
-      users    = dataframeFromCSV(sparkSession, usersPath)
-      salaries = dataframeFromCSV(sparkSession, salariesPath)
+    def salariesPath(sftpConfig: SFTPConfig): String =
+      s"sftp://${sftpConfig.sftpUser}:${sftpConfig.sftpPass}@${sftpConfig.sftpHost}" + s":${sftpConfig.sftpSalaryPath}"
 
-      // Sample operations to persist and query the Hive database
-      _                                        = HiveUserData.persistUserData(sparkSession, users, salaries)
-      (userDataFromHive, salariesDataFromHive) = HiveUserData.readUserData(sparkSession)
+    def readDataFramesFromSFTP(
+        sparkSession: SparkSession,
+        sftpConfig: SFTPConfig
+    ): IO[(DataFrame, DataFrame)] =
+      IO(
+        (
+          dataframeFromCSV(sparkSession, userPath(sftpConfig)),
```

Comment:
Same here, `dataframeFromCSV` seems like might contain the _effect_

Reply:
Changed in: a78818e